### PR TITLE
add user-api test

### PR DESCRIPTION
**Pull request by coderzc**

### CODE REVIEW

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            String user1 = "{\"user_password\":\"p1\"," +
+                           "\"user_email\":\"user1@baidu.com\"," +
+                           "\"user_phone\":\"111111\"," +
+                           "\"user_avatar\":\"image1" +
+                           ".jpg\"}";
+            Response r = client().put(path, (String) user.get("id"), user1,
+                                      ImmutableMap.of());
+            assertResponseStatus(200, r);
+        }
+
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:140]: remove the empty line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            if (user.get("user_name").equals("admin")) {
+                continue;
+            }
+            Response r = client().delete(path, (String) user.get("id"));
+        }
+        Response r = client().delete(path, "test1");
+        String result = assertResponseStatus(400, r);
+        String message = assertJsonContains(result, "message");
+        Assert.assertThat(message, CoreMatchers.containsString(
+                "Invalid user id:"));
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java]: align

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    protected void createUser(String name) {
+        String user = "{\"user_name\":\"" + name + "\",\"user_password\":\"p1" +
+                      "\", \"user_email\":\"user1@baidu.com\", " +
+                      "\"user_phone\"123456789\", \"user_avatar\":\"image1" +
+                      ".jpg\"}";
+        this.client().post(path, user);
+    }
+
+    protected List<Map<String, Object>> getList() {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:171]: listUsers

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        assertJsonContains(result2, "user_name");
+        assertJsonContains(result2, "user_password");
+        assertJsonContains(result2, "user_email");
+        assertJsonContains(result2, "user_phone");
+        assertJsonContains(result2, "user_avatar");
+
+        Response r3 = client().post(path, "{}");
+        assertResponseStatus(400, r3);
+
+
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:89]: remove empty line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+import com.baidu.hugegraph.util.JsonUtil;
+import com.google.common.collect.ImmutableMap;
+
+public class UserApiTest extends BaseApiTest {
+
+    private static final String path = "graphs/hugegraph/auth/users";
+
+    @After
+    public void clearUser() {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:41]: teardown

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        String user = "{\"user_name\":\"" + name + "\",\"user_password\":\"p1" +
+                      "\", \"user_email\":\"user1@baidu.com\", " +
+                      "\"user_phone\"123456789\", \"user_avatar\":\"image1" +
+                      ".jpg\"}";
+        this.client().post(path, user);
+    }
+
+    protected List<Map<String, Object>> getList() {
+        Response r = this.client().get(path,
+                                       ImmutableMap.of("limit", -1));
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:173]: define params var

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        String result = assertResponseStatus(200, r);
+        Map<String, List<Map<String, Object>>> resultMap = JsonUtil.fromJson(
+                result,
+                new TypeReference<Map<String, List<Map<String, Object>>>>() {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java]: define local var

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+                });
+        return resultMap.get("users");
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:179]: JsonUtil.fromJson().get("users");

File: hugegraph-dist/src/assembly/travis/run-api-test.sh [hunk 9]
```diff
@@ -5,9 +5,30 @@ set -ev
 TRAVIS_DIR=`dirname $0`
 VERSION=`mvn help:evaluate -Dexpression=project.version -q -DforceStdout`
 SERVER_DIR=hugegraph-$VERSION
+CONF=$TRAVIS_DIR/conf/hugegraph.properties
+REST_CONF=$TRAVIS_DIR/conf/rest-server.properties
```
Comment (javeme) [inline hugegraph-dist/src/assembly/travis/run-api-test.sh:9]: REST_SERVER_CONF

File: hugegraph-dist/src/assembly/travis/run-api-test.sh [hunk 10]
```diff
@@ -5,9 +5,30 @@ set -ev
...
+GREMLIN_CONF=$TRAVIS_DIR/conf/gremlin-server.yaml
 
 mvn package -DskipTests
+
+cp $CONF "${CONF}_bak"
```
Comment (javeme) [inline hugegraph-dist/src/assembly/travis/run-api-test.sh:14]: don't need to copy bak file

File: hugegraph-dist/src/assembly/travis/run-api-test.sh [hunk 11]
```diff
@@ -5,9 +5,30 @@ set -ev
 TRAVIS_DIR=`dirname $0`
 VERSION=`mvn help:evaluate -Dexpression=project.version -q -DforceStdout`
 SERVER_DIR=hugegraph-$VERSION
+CONF=$TRAVIS_DIR/conf/hugegraph.properties
+REST_CONF=$TRAVIS_DIR/conf/rest-server.properties
+GREMLIN_CONF=$TRAVIS_DIR/conf/gremlin-server.yaml
```
Comment (javeme) [inline hugegraph-dist/src/assembly/travis/run-api-test.sh:10]: GREMLIN_SERVER_CONF

File: hugegraph-dist/src/assembly/travis/run-api-test.sh [hunk 12]
```diff
@@ -5,9 +5,30 @@ set -ev
...
+}" >> $GREMLIN_CONF
+
 $TRAVIS_DIR/start-server.sh $SERVER_DIR || (cat $SERVER_DIR/logs/hugegraph-server.log && exit 1)
 mvn test -P api-test,$BACKEND || (cat $SERVER_DIR/logs/hugegraph-server.log && exit 1)
 $TRAVIS_DIR/build-report.sh
 $TRAVIS_DIR/stop-server.sh
+
+mv "${CONF}_bak" $CONF
+mv "${REST_CONF}_bak" $REST_CONF
+mv "${GREMLIN_CONF}_bak" $GREMLIN_CONF
```
Comment (javeme) [inline hugegraph-dist/src/assembly/travis/run-api-test.sh:34]: add empty line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 13]
```diff
@@ -36,9 +36,12 @@
 public class UserApiTest extends BaseApiTest {
 
     private static final String path = "graphs/hugegraph/auth/users";
+    private static final Map<String, Object> NO_LIMIT = ImmutableMap.of("limit",
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:39]: just define NO_LIMIT = -1;

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 14]
```diff
@@ -168,14 +169,12 @@ protected void createUser(String name) {
...
+    protected List<Map<String, Object>> listUsers() {
+        Response r = this.client().get(path, NO_LIMIT);
         String result = assertResponseStatus(200, r);
-        Map<String, List<Map<String, Object>>> resultMap = JsonUtil.fromJson(
-                result,
-                new TypeReference<Map<String, List<Map<String, Object>>>>() {
-                });
+        Map<String, List<Map<String, Object>>> resultMap =
+                JsonUtil.fromJson(result, new TypeReference<Map<String,
+                                  List<Map<String, Object>>>>() {});
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:177]: return JsonUtil.fromJson(result, type).get("users");

Comment (coderzc) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:177]: it need to cast

Comment (javeme) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:177]: update alignment

File: hugegraph-dist/src/assembly/travis/run-api-test.sh [hunk 15]
```diff
@@ -5,8 +5,21 @@ set -ev
...
 VERSION=`mvn help:evaluate -Dexpression=project.version -q -DforceStdout`
 SERVER_DIR=hugegraph-$VERSION
+CONF=$SERVER_DIR/conf/hugegraph.properties
+REST_SERVER_CONF=$SERVER_DIR/conf/rest-server.properties
+GREMLIN_SERVER_CONF=$SERVER_DIR/conf/gremlin-server.yaml
 
 mvn package -DskipTests
+
+sed -i 's/#auth.authenticator=/auth.authenticator=com.baidu.hugegraph.auth.StandardAuthenticator/' $REST_SERVER_CONF
+sed -i 's/gremlin.graph=.*/gremlin.graph=com.baidu.hugegraph.auth.HugeFactoryAuthProxy/' $CONF
```
Comment (javeme) [inline hugegraph-dist/src/assembly/travis/run-api-test.sh:15]: seem no this option

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 16]
```diff
@@ -162,19 +162,20 @@ public void testDelete() {
 
     protected void createUser(String name) {
         String user = "{\"user_name\":\"" + name + "\",\"user_password\":\"p1" +
-                      "\", \"user_email\":\"user1@baidu.com\", " +
-                      "\"user_phone\"123456789\", \"user_avatar\":\"image1" +
+                      "\", \"user_email\":\"user1@baidu.com\"," +
+                      "\"user_phone\":123456789,\"user_avatar\":\"image1" +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:166]: keep string

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/GremlinApiTest.java [hunk 17]
```diff
@@ -93,7 +93,13 @@ public void testClearAndInit() {
         assertResponseStatus(200, client().post(path, body));
 
         body = "{"
-                + "\"gremlin\":\"hugegraph.initBackend()\","
+                + "\"gremlin\":\"import com.baidu.hugegraph.auth.HugeUser;"
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/GremlinApiTest.java:96]: can't separate clearBackend and initBackend into 2 parts if enabled auth

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/GremlinApiTest.java [hunk 18]
```diff
@@ -86,14 +86,14 @@ public void testScript() {
...
-                + "\"aliases\":{\"g\":\"__g_hugegraph\"}}";
-        assertResponseStatus(200, client().post(path, body));
-
-        body = "{"
-                + "\"gremlin\":\"hugegraph.initBackend()\","
+                + "\"gremlin\":\"import com.baidu.hugegraph.auth.HugeUser;"
+                + "import com.baidu.hugegraph.util.StringEncoding;"
+                + "hugegraph.clearBackend();"
+                + "hugegraph.initBackend();"
+                + "HugeUser admin = new HugeUser('admin');"
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/GremlinApiTest.java:93]:
```
auth = hugegraph.authManager();
admin = auth.findUser('admin');
hugegraph.clearBackend();
hugegraph.initBackend();
auth.createUser(admin);
```

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java [hunk 19]
```diff
@@ -52,6 +53,8 @@
 
     private static String BASE_URL = "http://127.0.0.1:8080";
     private static String GRAPH = "hugegraph";
+    protected static final String USERNAME = "admin";
+    protected static final String PASSWORD = "pa";
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/BaseApiTest.java:57]: set private

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+        Response r = this.client().post(path, user);
+        assertResponseStatus(201, r);
+    }
+
+    protected List<Map<String, Object>> listUsers() {
+        Response r = this.client().get(path, ImmutableMap.of("limit",
+                                                             NO_LIMIT));
+        String result = assertResponseStatus(200, r);
+        Map<String, List<Map<String, Object>>> resultMap =
+        JsonUtil.fromJson(result, new TypeReference<Map<String,
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/UserApiTest.java:177]: improve align

File: hugegraph-dist/src/assembly/travis/run-api-test.sh [hunk 21]
```diff
@@ -5,8 +5,21 @@ set -ev
 TRAVIS_DIR=`dirname $0`
 VERSION=`mvn help:evaluate -Dexpression=project.version -q -DforceStdout`
 SERVER_DIR=hugegraph-$VERSION
+CONF=$SERVER_DIR/conf/hugegraph.properties
+REST_SERVER_CONF=$SERVER_DIR/conf/rest-server.properties
+GREMLIN_SERVER_CONF=$SERVER_DIR/conf/gremlin-server.yaml
 
 mvn package -DskipTests
+
+sed -i 's/#auth.authenticator=/auth.authenticator=com.baidu.hugegraph.auth.StandardAuthenticator/' $REST_SERVER_CONF
```
Comment (javeme) [inline hugegraph-dist/src/assembly/travis/run-api-test.sh:14]: set admin_token

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/GremlinApiTest.java (hunk 18 above)
Comment (coderzc) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/api/GremlinApiTest.java:93]: image

File: hugegraph-test/src/main/java/com/baidu/hugegraph/api/GremlinApiTest.java [hunk 22]
```diff
@@ -88,11 +88,10 @@ public void testClearAndInit() {
...
                 + "\"gremlin\":\"import com.baidu.hugegraph.auth.HugeUser;"
                 + "import com.baidu.hugegraph.util.StringEncoding;"
+                + "def auth = hugegraph.authManager();"
+                + "def admin = auth.findUser('admin');"
                 + "hugegraph.clearBackend();"
                 + "hugegraph.initBackend();"
-                + "HugeUser admin = new HugeUser('admin');"
-                + "admin.password(StringEncoding.hashPassword('pa'));"
-                + "admin.creator('system');"
                 + "hugegraph.hugegraph().authManager().createUser(admin);\","
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/api/GremlinApiTest.java:95]: auth.createUser(admin);

File: hugegraph-dist/src/assembly/travis/run-api-test.sh (hunk 15 above)
Comment (javeme) [reply hugegraph-dist/src/assembly/travis/run-api-test.sh:16]: seem no the option named "gremlin.graph"

**Review by javeme — APPROVED** [review]

**Review by houzhizhen — APPROVED** [review]

**Pull request merged**